File: java/completed-affine-cipher/src/main/java/AffineCipher.java
/*

Since this exercise has a difficulty of > 4 it doesn't come
with any starter implementation.
This is so that you get to practice creating classes and methods
which is an important part of programming in Java.

Please remove this comment when submitting your solution.

*/
public class AffineCipher {
    private static final int a = 'a';
    private static final int m = 26;

    public String encode(String text, int a, int b) {
        if (gcd(a, m) != 1)
            throw new IllegalArgumentException("Error: keyA and alphabet size must be coprime.");
        text = text.toLowerCase();
        StringBuilder encoded = new StringBuilder("");
        for (int i = 0, k = 1; i < text.length(); i++) {
            char x = text.charAt(i);
            if (Character.isAlphabetic(x)) {
                encoded.append((char) ((a * ((int) x - AffineCipher.a) + b) % m + AffineCipher.a));
                if (k % 5 == 0) encoded.append(' ');
                k += 1;
            } else if (Character.isDigit(x)) {
                encoded.append(x);
                if (k % 5 == 0) encoded.append(' ');
                k += 1;
            }
        }
        return encoded.toString().trim();
    }

    public String decode(String text, int a, int b) {
        if (gcd(a, 26) != 1)
            throw new IllegalArgumentException("Error: keyA and alphabet size must be coprime.");
        text = text.toLowerCase();
        StringBuilder decoded = new StringBuilder("");
        for (int i = 0; i < text.length(); i++) {
            char y = text.charAt(i);
            if (Character.isAlphabetic(y)) {
                decoded.append((char) (absm(mmi(a, m) * (y - AffineCipher.a - b), m) + AffineCipher.a));
            } else if (Character.isDigit(y)) {
                decoded.append(y);
            }
        }
        return decoded.toString().trim();
    }

    private static int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    private static int mmi(int a, int b) {
        for (int i = 1; i < b; i++)
            if ((a * i) % b == 1)
                return i;
        return 1;
    }

    public static int absm(int n, int m) {
        while (n < 0) n += m;
        while (n >= m) n -= m;
        return n;
    }
}

File: javascript/completed-affine-cipher/affine-cipher.js
let a_code = 'a'.charCodeAt(0)

function gcd(a, b) {
  return !b ? a : gcd(b, a % b)
}

function check(a) {
  if (gcd(a, 26) !== 1)
    throw new Error('a and m must be coprime.')
}

function isAlpha(c) {
  let z = 'z'.charCodeAt(0)
  return a_code <= c.charCodeAt(0) && c.charCodeAt(0) <= z;
}

function isDigit(c) {
  let zero = '0'.charCodeAt(0)
  let nine = '9'.charCodeAt(0)
  return zero <= c.charCodeAt(0) && c.charCodeAt(0) <= nine;
}

export const encode = (phrase, key) => {
  phrase = phrase.toLowerCase()
  const { a, b } = key
  check(a)
  let encoded = ""
  const addSymbol = (code) => {
    encoded += String.fromCharCode([code])
  }
  for (let i = 0, k = 0; i < phrase.length; i++) {
    if (isAlpha(phrase.charAt(i))) {
      addSymbol((a * (phrase.charCodeAt(i) - a_code) + b) % 26 + a_code)
      k += 1
      if (k === 5) {
        encoded += " "
        k = 0
      }
    } else if (isDigit(phrase.charAt(i))) {
      encoded += phrase.charAt(i)
      k += 1
      if (k === 5) {
        encoded += " "
        k = 0
      }
    }
  }
  if (encoded[encoded.length - 1] === ' ')
    return encoded.substring(0, encoded.length - 1)
  return encoded;
};

function mmi(a) {
  for (let i = 1; i < 26; i++) {
    if (a * i % 26 === 1) return i
  }
}

export const decode = (phrase, key) => {
  phrase = phrase.toLowerCase()
  let { a, b } = key
  check(a)
  let decoded = ""
  a = mmi(a)
  for (const c of phrase) {
    if (isAlpha(c)) {
      let code = c.charCodeAt(0) - a_code - b
      while (code < 0) code += 26
      decoded += String.fromCharCode([a * code % 26 + a_code])
    }
    else if (isDigit(c))
      decoded += c
  }
  return decoded
};
